## scDeepHIM:
1. Read count matrices for histone modifications for each samples in training groups were generated using either the bamCoverage function or an alternative method. These matrices were partitioned into input files (H3K27ac, H3K27me3, H3K9me3) and target files (H3K4me1, H3K4me3, H3K36me3). We trained scDeepHIM models with profiles of six histone modifications using 80% of the samples and evaluated its performance on the remaining 20%. During the training phase, the scDeepHIM model optimized an objective function to predict H3K4me1, H3K4me3, and H3K36me3 signals using the provided H3K27ac, H3K27me3, and H3K9me3 signals from both bulk ChIP-seq profiles and single-cell ChIP-seq profiles.
2. Following training, scDeepHIM can output 3*273119 matrices containing predicted signals for three histone modifications (H3K4me1, H3K4me3, and H3K36me3) at 273119 genomic regions when provided with a 3*273119 matrix containing H3K27ac, H3K27me3, and H3K9me3 signals. These matrices should be transformed into formatted read count matrices containing four columns (chr, start, end, signal).
3. Predicted read count matrices were then converted to bedGraph format and utilized for subsequent analyses.
